### ISSUE TITLE
`accelerate.gather` returning incorrect number of items

### ISSUE THREAD

**Issue by kheyer** [huggingface/accelerate#2014]
### System Info

```Shell
- `Accelerate` version: 0.23.0
- Platform: Linux-5.15.0-84-generic-x86_64-with-glibc2.31
- Python version: 3.9.18
- Numpy version: 1.26.0
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 125.65 GB
- GPU type: NVIDIA GeForce RTX 3090
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: False
        - main_training_function: main
        - downcast_bf16: False
        - tpu_use_cluster: False
        - tpu_use_sudo: False
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I'm using accelerate to compute embeddings on multiple GPUs. When trying to get the final output using `accelerate.gather`, the total number of items returned is incorrect.

This is the script I'm using:

```python
from tqdm import tqdm
import datasets
from datasets import disable_caching
disable_caching()
from accelerate import Accelerator

import torch
from torch.utils.data import DataLoader

from transformers import RobertaTokenizerFast, RobertaForMaskedLM, DataCollatorWithPadding

model_name = ...
embed_batch_size = 1024

tokenizer = RobertaTokenizerFast.from_pretrained(model_name, max_len=256)
collator = DataCollatorWithPadding(tokenizer, padding=True, return_tensors='pt')


def main(dataset):
    embeddings = []
    accelerator = Accelerator()
    
    with accelerator.main_process_first():
        model = RobertaForMaskedLM.from_pretrained(model_name)
    
    dataloader = DataLoader(dataset, batch_size=embed_batch_size, collate_fn=collator, drop_last=False)
    
    model, dataloader = accelerator.prepare(model, dataloader)
    model.eval()
    
    with torch.no_grad():
        for batch in tqdm(dataloader):
            unwrap_model = accelerator.unwrap_model(model)
            outputs = unwrap_model(**batch, output_hidden_states=True)

            full_embeddings = outputs[-1][-1]
            mask = batch['attention_mask']
            mask = mask.to(full_embeddings.device)
            mean_embeddings = ((full_embeddings * mask.unsqueeze(-1)).sum(1) / mask.sum(-1).unsqueeze(-1))
            embeddings += accelerator.gather(mean_embeddings).detach().cpu()
                        
    return embeddings
    
if __name__ == "__main__":
    # dataset is processed, contains only `input_ids` and `attention_mask`
    dataset = datasets.load_from_disk(dataset_filename)
    embeddings = main(dataset)
    print(len(embeddings))
    print(len(dataset))
```

### Expected behavior

When running with `accelerate launch`, the script prints `456704` for the number of embeddings returned, and the `456364` for the length of the dataset. I assume data points are being added somewhere to pad out the final batch, but how do I know which items are padding? I need to associate the computed embeddings 1:1 with the inputs. Is there a way to do this?
